mangaki/mangaki/utils/mal.py

```diff
@@ -31,149 +40,256 @@ def random_ip():
+
+        if '-00' in xml_date:
+            return xml_date.replace('-00', '-01')
+
+        return xml_date
+
+    @property
+    def source_url(self) -> str:
+        return (
+            'https://myanimelist.net/{}/{}'.format(self.work_type.value, self.raw_properties['id'])
+        )
 
-    SEARCH_URL = 'http://myanimelist.net/api/anime/search.xml'
+
+class MALClient:
+    SEARCH_URL = 'https://myanimelist.net/api/{type}/search.xml'
     HEADERS = {
-        'X-Real-IP': random_ip(),
-        'User-Agent': settings.MAL_USER_AGENT
+        'User-Agent': getattr(settings, 'MAL_USER_AGENT', 'mangaki')
     }
 
-    r = requests.get(SEARCH_URL, params={'q': query}, headers=HEADERS,
-                     auth=(settings.MAL_USER, settings.MAL_PASS))
-    html_code = html.unescape(re.sub(r'&amp;([A-Za-z]+);', r'&\1;', r.text))
-    xml = re.sub(r'&([^alg])', r'&amp;\1', _encoding_translation(html_code))
+    def __init__(self):
+        if not hasattr(settings, 'MAL_USER') or not hasattr(settings, 'MAL_PASS'):
+            self.is_available = False
+        else:
+            self.session = requests.Session()
+            self.session.headers = self.HEADERS
+            self.session.auth = (settings.MAL_USER, settings.MAL_PASS)
+            self.is_available = True
 
-    entries = []
-    try:
+    @staticmethod
+    def _translate_http_exceptions(resp: requests.Response) -> None:
+        if resp.status_code == requests.codes['FORBIDDEN']:
+            raise RuntimeError('Invalid MAL credentials!')
+
+        if not resp.status_code == requests.codes['ALL_GOOD']:
+            raise RuntimeError('MAL request failure!')
+
+    def list_works_from_a_user(self, work_type: MALWorks, username: str) -> str:
+        resp = (
+            self.session.get(
+                'https://myanimelist.net/malappinfo.php?u={}&status=all&type={}'.format(username, work_type.value)
+            )
+        )
+
+        return resp.text
+
+    def _search_api(self, work_type: MALWorks, query: str) -> requests.Response:
+        return self.session.get(
+            self.SEARCH_URL.format(type=work_type.value),
+            params={
+                'q': query
+            })
+
+    def search_work(self, work_type: MALWorks, query: str) -> MALEntry:
+        resp = self._search_api(work_type, query)
+        self._translate_http_exceptions(resp)
+
+        html_code = html.unescape(re.sub(r'&amp;([A-Za-z]+);', r'a\1;', resp.text))
+        xml = re.sub(r'&([^alg])', r'&amp;\1', _encoding_translation(html_code))
+
+        return MALEntry(ET.fromstring(xml).find('entry'), work_type)
+
+    def search_works(self, work_type: MALWorks, query: str) -> List[MALEntry]:
+        resp = self._search_api(work_type, query)
+        self._translate_http_exceptions(resp)
+
+        html_code = html.unescape(re.sub(r'&amp;([A-Za-z]+);', r'a\1;', resp.text))
+        xml = re.sub(r'&([^alg])', r'&amp;\1', _encoding_translation(html_code))
```

Comment:
Right, I put the work to deduplicate the code, but… forgot the last pass :D. Thanks!